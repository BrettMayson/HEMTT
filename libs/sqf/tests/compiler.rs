#![allow(clippy::unwrap_used)]

pub use float_ord::FloatOrd as Scalar;
use hemtt_preprocessor::Processor;
use hemtt_sqf::{compiler::serializer::Compiled, parser::database::Database};
use hemtt_workspace::LayerType;

macro_rules! compile {
    ($dir: expr, $file:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $file>]() {
                let (bin, compiled) = optimize($dir, stringify!($file));
                // bin to hex
                let hex = bin.iter().map(|b| format!("{:02x}", b)).collect::<String>();
                insta::assert_snapshot!(hex);
                insta::assert_debug_snapshot!(compiled);
            }
        }
    };
}

compile!("optimizer", consume_array);
compile!("optimizer", static_math);
compile!("optimizer", scalar);
compile!("optimizer", string_case);
compile!("simple", dev);
compile!("simple", eventhandler);
compile!("simple", foreach);
compile!("simple", format_font);
compile!("simple", get_visibility);
compile!("simple", hash_select);
compile!("simple", hello);
compile!("simple", include);
compile!("simple", oneline);
compile!("simple", semicolons);

const ROOT: &str = "tests/";

fn optimize(folder: &str, file: &str) -> (Vec<u8>, Compiled) {
    let folder = std::path::PathBuf::from(ROOT).join(folder);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join(format!("{file}.sqf")).unwrap();
    let processed = Processor::run(&source).unwrap();
    let mut writer = Vec::new();
    let statements = hemtt_sqf::parser::run(&Database::a3(false), &processed).unwrap();
    let compiled = statements.compile(&processed).unwrap();
    statements
        .optimize()
        .compile_to_writer(&processed, &mut writer)
        .unwrap();
    (writer, compiled)
}
