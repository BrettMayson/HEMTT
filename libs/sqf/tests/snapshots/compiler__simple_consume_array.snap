---
source: libs/sqf/tests/compiler.rs
expression: hex
---
01000000059c0000000200261100060000706172616d730100002b1000006d697373696f6e6e616d6573706163650b00006765747661726961626c650100007a0b0000732b4400004570090000766563746f72616464150000706f736974696f6e63616d657261746f776f726c640200005f7806000072616e646f6d04000074696d650100003e02000069660400007468656e0200005f7a050000706172616d1100000158030000020a13000402000000010200005f61940001620403002d410002e00204630400000000049c01060100006102000080bfbc050101000062960303019c0369033f6105009c03020200401c4698012a260000806a04204160030240020000407100809100bf6d02c088090700140100000800008008740f00060107000001090000010a0000010b00002601000006000080900f06010c0000010d000401e5166428280000266501510100236c696e65203120225c636f6e73756d655f61727261792e737166220a706172616d73205b225f61222c20225f62225d3b0a0a3055002c68040463222c205b5d5d6404000d6d697373696f6e4e616d657370616365206765745661726961626c65205b2272092d3163057a20732b6500628c08032c20747275657c0300245b312c305d20766563746f7241646420703b0a0a706f736974696f6e43616d657261546f576f726c64205b31303030302c20302c2031640174070d0a72616e646f6d205b302c205f782c20880e002c70726976617465205f7a203d206966202874696d65203e20313029207468656e207b20313b323b333b343b207d20656c7365207b202d313b2d323b207d3bd41f7e1d64226f0c5b5d209c21650265660200117032010000002c7c0002010000021dc001040000010100023b740103030001000233a4010200000004556b010500027e4e0366b00102030000077e68010607000400010300028db0010101000380a4010505000001040007ae7401030900060003a4a401604201050002b275010b60450201060007e87401060e00090001070008e5b00101030002dda5010a745870440610000b000108000307b001010c0002fe64040610000d00010e00030db802010e0006f1b802050f0000010f0002317f021200109c52000200000000000000011000034301000000140010001100000201001200002f636f6e73756d655f61727261792e737166031200000000000000
